core/src/main/scala/org/apache/spark/util/AccumulatorV2.scala

```diff
@@ -136,15 +179,76 @@ abstract class AccumulatorV2[IN, OUT] extends Serializable {
   def reset(): Unit
 
   /**
+   * Takes the inputs and accumulates. e.g. it can be a simple `+=` for counter accumulator.
+   * Developers should extend addImpl to customize the adding functionality.
+   */
+  final def add(v: IN): Unit = {
+    if (metadata != null && metadata.dataProperty) {
+      dataPropertyAdd(v)
+    } else {
+      addImpl(v)
+    }
+  }
+
+  private def dataPropertyAdd(v: IN): Unit = {
+    // Add first for localValue & AccumulatorInfo
+    addImpl(v)
+    if (metadata != null && metadata.dataProperty) {
+      val updateInfo = TaskContext.get().getRDDPartitionInfo()
+      val base = pending.getOrElse(updateInfo, copyAndReset())
+      base.atDriverSide = false
```

Comment:
So the problem is that we are constructing new instances of the underlying accumulator for new partitions, which by default has `atDriverSide` set to true so we need to set it to false because we are not at the driver side. I'll add a comment for this as well :)